### PR TITLE
server/balancer: guarantee replicas are safe if possible

### PR DESCRIPTION
**Pull request by huachaohuang**
We can not do balance safely if we have only 3 data centers, because of https://github.com/pingcap/tikv/issues/1468.
So, the current implementation guarantee safety if we have more than 3 data centers, but we will still balance if we have only 3 data cetners.

### CODE REVIEW

File: server/balancer.go [hunk 1]
```diff
@@ -181,33 +182,35 @@ func (r *replicaChecker) Check(region *regionInfo) Operator {
...
 
-	return r.checkBetterPeer(region)
+	return r.checkBestReplacement(region)
 }
 
 // selectBestPeer returns the best peer in other stores.
 func (r *replicaChecker) selectBestPeer(region *regionInfo, filters ...Filter) (*metapb.Peer, float64) {
-	filters = append(filters, r.filters...)
+	// Add some must have filters.
+	filters = append(filters, newStateFilter(r.opt))
```
Comment (overvenus) [inline server/balancer.go]: How about adding a `healthFilter`?

File: server/balancer_test.go [hunk 2]
```diff
@@ -247,31 +247,86 @@ func (s *testStorageBalancerSuite) TestReplicaScore(c *C) {
...
 	checkTransferPeer(c, sb.Schedule(cluster), 1, 6)
 
-	// Take down 4,5,6
+	// Store 7 has the same storage ratio with store 6, but in a different host.
+	tc.addLabelsStore(7, 1, 0.2, map[string]string{"zone": "z1", "rack": "r1", "host": "h2"})
+	checkTransferPeer(c, sb.Schedule(cluster), 1, 7)
+
+	// If store 7 is not available, we wait.
+	tc.setStoreDown(7)
+	c.Assert(sb.Schedule(cluster), IsNil)
```
Comment (huachaohuang) [inline server/balancer_test.go]: It's not safe to balance 1->6, we don't need to balance immediately if 7 is down.

File: server/balancer_test.go [hunk 3]
```diff
@@ -247,31 +247,86 @@ func (s *testStorageBalancerSuite) TestReplicaScore(c *C) {
...
+	c.Assert(sb.cache.get(1), IsTrue)
```
Comment (huachaohuang) [inline server/balancer_test.go]: Store 1 is not able to balance temporarily when 7 is down, we can skip it for a while.

**Comment by huachaohuang** [comment]
@overvenus Yes, the option has been removed for simplicity.

File: server/balancer.go [hunk 4]
```diff
@@ -181,33 +182,35 @@ func (r *replicaChecker) Check(region *regionInfo) Operator {
...
-		score := r.rep.GetReplicaScore(regionStores, store)
+		score := r.rep.GetDistinctScore(stores, store)
 		if bestStore == nil || compareStoreScore(store, score, bestStore, bestScore) > 0 {
 			bestStore = store
 			bestScore = score
 		}
 	}
 
-	if bestStore == nil {
+	if bestStore == nil || filterTarget(bestStore, r.filters) {
```
Comment (huachaohuang) [inline server/balancer.go]: `r.filters` are used to filter temporarily unavailable stores. We may choose an unsafe store if we filter those temporarily unavailable stores before calculating the score.

File: server/balancer.go (hunk 1 above)
Comment (huachaohuang) [inline server/balancer.go]: `healthFilter` will be used after the best store has been found, or the best store can be filtered if it is temporarily unavailable.

File: server/balancer.go (hunk 4 above)
Comment (disksing) [inline server/balancer.go]: I see.

**Comment by disksing** [comment]
LGTM

File: server/store.go [hunk 5]
```diff
@@ -107,6 +107,18 @@ func (s *storeInfo) getLabelValue(key string) string {
...
 }
 
+func (s *storeInfo) getLocationID(keys []string) string {
+	id := ""
+	for _, k := range keys {
+		v := s.getLabelValue(k)
+		if len(v) == 0 {
+			return ""
+		}
+		id += s.getLabelValue(k)
```
Comment (siddontang) [inline server/store.go]: can we use v here?

File: server/store.go [hunk 6]
```diff
@@ -107,6 +107,18 @@ func (s *storeInfo) getLabelValue(key string) string {
 	return ""
 }
 
+func (s *storeInfo) getLocationID(keys []string) string {
+	id := ""
+	for _, k := range keys {
+		v := s.getLabelValue(k)
+		if len(v) == 0 {
+			return ""
```
Comment (siddontang) [inline server/store.go]: no need to consider the latter key which value is not empty?

Comment (huachaohuang) [inline server/store.go]: I don't think we should support missing label cases, it's error prone and hard to explain the behavior.
If we want to update the configurations or labels, we can just turn off the balance temporarily.

File: server/store.go (hunk 5 above)
Comment (huachaohuang) [inline server/store.go]: I concat the values into an id, so id will be more appropriate.

Comment (siddontang) [inline server/store.go]: what I mean is `id += v`, no need to call `s.getLabelValue(k)` twice.

File: server/store.go (hunk 6 above)
Comment (siddontang) [inline server/store.go]: So we should forbid the missing label in TiKV, and if TiKV reports the label without a value, we should log an error and don't add this label too.

**Comment by siddontang** [comment]
Rest LGTM

File: server/cluster.go [hunk 7]
```diff
@@ -325,15 +325,24 @@ func (c *RaftCluster) putStore(store *metapb.Store) error {
...
 		s.Address = store.Address
 		s.Labels = store.Labels
-		return cluster.putStore(s)
 	}
 
-	// Store does not exist, add a new store.
-	return cluster.putStore(newStoreInfo(store))
+	// Check location labels.
+	for _, k := range c.s.cfg.Replication.LocationLabels {
+		if v := s.getLabelValue(k); len(v) == 0 {
```
Comment (siddontang) [inline server/cluster.go]: should check before update the store info?

Comment (huachaohuang) [inline server/cluster.go]: That's ok, it is a clone.

Comment (siddontang) [inline server/cluster.go]: LGTM

**Pull request merged**